File: frontend/src/api/events.ts
/**
 * The live run stream.
 *
 * The backend's contract, which this file exists to honour rather than reinvent:
 *
 * * every event carries a `seq`, monotonic within a run, starting at 1;
 * * reconnecting with `since_seq=<last rendered>` replays everything after it;
 * * if the client slept longer than the ring buffer, the server sends a `gap` frame FIRST, saying how
 *   many events are gone forever.
 *
 * That last one is the reason this is a hand-written socket and not a two-line `new WebSocket`. A UI
 * that silently renders a short replay draws a run that never had those steps -- an operator reading
 * it would conclude the cell skipped a stage it actually performed. So `gap` is surfaced as an event
 * in the list, visibly, and never swallowed.
 *
 * `keepalive` frames are the opposite: they carry no run information and exist only to prove the
 * socket is alive. They are consumed here and turned into a timestamp the UI can show as "last heard
 * from the cell", never into a row.
 */

export type Severity = 'info' | 'warn' | 'error' | 'success'

/** One thing that happened, addressed to both a person and a program. */
export interface RunEvent {
  type: string
  run_id: string
  seq: number
  ts: number
  severity: Severity
  /** A plain sentence, written by the backend. The console never composes its own. */
  human: string
  step: string
  step_index: number | null
  step_total: number | null
  /** The machine payload. Never a stringified version of `human`. */
  data: Record<string, unknown>
}

/**
 * A hole in the record, rendered as loudly as anything else.
 *
 * Synthesised locally from the server's `gap` frame so the list has one element type. `dropped` is the
 * server's count, not a guess.
 */
export interface GapEvent extends RunEvent {
  type: 'gap'
  dropped: number
}

export type StreamState = 'connecting' | 'live' | 'closed' | 'failed'

export interface StreamHandlers {
  onEvent: (event: RunEvent) => void
  onState: (state: StreamState, detail?: string) => void
  /** Fired on every frame including keepalives, so a UI can show liveness without inventing rows. */
  onHeartbeat?: (at: number) => void
}

function socketUrl(runId: string, sinceSeq: number): string {
  const scheme = window.location.protocol === 'https:' ? 'wss:' : 'ws:'
  return `${scheme}//${window.location.host}/v1/events?run_id=${encodeURIComponent(runId)}&since_seq=${sinceSeq}`
}

/**
 * Follow one run until it is closed or the caller stops it.
 *
 * Reconnects automatically, resuming from the last `seq` actually delivered to `onEvent` -- not from
 * the last one received, and not from zero. Resuming from zero would replay the whole run into a list
 * that already holds it; resuming from a seq that was received but not rendered would lose a row.
 *
 * Returns a stop function. Calling it is idempotent and suppresses reconnection.
 */
export function followRun(runId: string, handlers: StreamHandlers): () => void {
  let cursor = 0
  let stopped = false
  let socket: WebSocket | null = null
  let retry: ReturnType<typeof setTimeout> | null = null
  // Backs off so a backend that is down does not become a reconnect storm, but stays short enough
  // that an operator watching a bring-up does not think the console has given up.
  let backoffMs = 500

  const open = () => {
    if (stopped) return
    handlers.onState('connecting')
    let ws: WebSocket
    try {
      ws = new WebSocket(socketUrl(runId, cursor))
    } catch (err) {
      handlers.onState('failed', String(err))
      schedule()
      return
    }
    socket = ws

    ws.onopen = () => {
      backoffMs = 500
      handlers.onState('live')
    }

    ws.onmessage = (message) => {
      handlers.onHeartbeat?.(Date.now())
      let frame: Record<string, unknown>
      try {
        frame = JSON.parse(message.data as string)
      } catch {
        return // an unparseable frame is a backend bug; dropping it beats crashing the console
      }
      const type = String(frame.type ?? '')
      if (type === 'keepalive') return
      if (type === 'gap') {
        // Not a run event -- it has no seq of its own -- so it is given the cursor it interrupts and
        // shown in place. It must never be filtered out as "not a real event".
        handlers.onEvent({
          type: 'gap',
          run_id: runId,
          seq: cursor,
          ts: Date.now() / 1000,
          severity: 'warn',
          human: String(frame.human ?? 'Events were lost before this point.'),
          step: '',
          step_index: null,
          step_total: null,
          data: { dropped: frame.dropped ?? null },
          dropped: Number(frame.dropped ?? 0),
        } as GapEvent)
        return
      }
      const event = frame as unknown as RunEvent
      if (typeof event.seq !== 'number') return
      // Guard against a duplicate replay after a reconnect: the server is asked for `> cursor`, but a
      // race at the boundary is cheaper to drop here than to de-duplicate in every screen.
      if (event.seq <= cursor) return
      cursor = event.seq
      handlers.onEvent(event)
    }

    ws.onerror = () => {
      // Deliberately quiet: `onclose` always follows, and reporting both makes one dropped connection
      // look like two failures in the UI.
    }

    ws.onclose = () => {
      socket = null
      if (stopped) {
        handlers.onState('closed')
        return
      }
      handlers.onState('closed')
      schedule()
    }
  }

  const schedule = () => {
    if (stopped || retry) return
    retry = setTimeout(() => {
      retry = null
      open()
    }, backoffMs)
    backoffMs = Math.min(backoffMs * 2, 5000)
  }

  open()

  return () => {
    stopped = true
    if (retry) clearTimeout(retry)
    retry = null
    socket?.close()
    socket = null
  }
}

File: api/events.py
"""The event stream, and the reason it has sequence numbers.

A run must survive the browser. An operator starts a pick, the laptop sleeps, the tab is closed, the
wifi drops. The arm does not care: it is holding a part. So the run finishes regardless, every event
it produced is kept, and a client that comes back asks for everything after seq 41 and receives 42
onwards before the live feed resumes. Without a monotonic ``seq`` there is no way to ask that
question, and a UI that reconnects can only guess whether it missed anything.

The buffer is bounded and says when it dropped. A ring buffer per run, and a reconnect asking for a
``since_seq`` older than the buffer gets a truthful gap marker instead of a silently short replay. A
console that renders "events 12-40 are gone" is annoying; one that renders 41 onwards as if nothing
were missing is wrong.

Two audiences, one envelope. ``human`` is a plain sentence for the operator; ``data`` is the machine
payload. Both, always: the person watching a demo reads one, the person diagnosing a failure reads the
other, and an envelope that carries only one of them forces the other to guess.
"""

from __future__ import annotations

import threading
import time
from collections import deque
from dataclasses import dataclass, field
from enum import StrEnum
from typing import Any

__all__ = ["EventEnvelope", "EventHub", "Severity"]

#: Events kept per run. At five stage events per attempt and five attempts per pick, this holds dozens
#: of picks: more than a browser needs to catch up after a sleep, and small enough that a forgotten
#: long-running server does not grow without bound.
_RING = 2048


class Severity(StrEnum):
    """How loudly to render. Not a log level: it is what the operator's eye should go to first."""

    INFO = "info"
    #: Something the operator should notice but that did not stop anything.
    WARN = "warn"
    #: The run stopped, or the cell refused.
    ERROR = "error"
    #: A milestone worth celebrating on a demo screen: a successful pick.
    SUCCESS = "success"


@dataclass(frozen=True, slots=True)
class EventEnvelope:
    """One thing that happened, addressed to both a person and a program."""

    type: str
    run_id: str
    #: Monotonic within a run, starting at 1. A reconnecting client resumes from the last one it saw.
    seq: int
    #: Unix seconds. Wall clock on purpose: it is compared against the operator's own clock.
    ts: float
    severity: Severity = Severity.INFO
    #: A plain sentence. "Ranked 3 candidates, best score 0.85."
    human: str = ""
    #: Where in the run this sits, when that is meaningful.
    step: str = ""
    step_index: int | None = None
    step_total: int | None = None
    #: The machine payload. Never a stringified version of ``human``.
    data: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        return {
            "type": self.type,
            "run_id": self.run_id,
            "seq": self.seq,
            "ts": self.ts,
            "severity": str(self.severity),
            "human": self.human,
            "step": self.step,
            "step_index": self.step_index,
            "step_total": self.step_total,
            "data": self.data,
        }


class EventHub:
    """Per-run sequence numbers, a bounded history, and live subscribers.

    Publishing is safe from the pick thread; subscribing is safe from a request thread. A run
    publishes from a robot-driving thread while a browser subscribes from an event loop, so every
    mutation is under one lock: an interleaving that gives two events the same ``seq`` leaves a
    reconnecting client unable to tell what it missed.
    """

    def __init__(self, *, capacity: int = _RING) -> None:
        self._capacity = capacity
        self._lock = threading.RLock()
        self._history: dict[str, deque[EventEnvelope]] = {}
        self._next_seq: dict[str, int] = {}
        #: One condition per hub; waiters re-check their own run's tail on every publish. Simpler than
        #: per-run conditions and correct: a wake-up meant for another run costs one comparison.
        self._published = threading.Condition(self._lock)

    def publish(
        self,
        run_id: str,
        event_type: str,
        *,
        human: str = "",
        severity: Severity = Severity.INFO,
        step: str = "",
        step_index: int | None = None,
        step_total: int | None = None,
        **data: Any,
    ) -> EventEnvelope:
        """Append one event and wake anything waiting. Never raises for an unknown run."""
        with self._lock:
            seq = self._next_seq.get(run_id, 1)
            self._next_seq[run_id] = seq + 1
            envelope = EventEnvelope(
                type=event_type, run_id=run_id, seq=seq, ts=time.time(),
                severity=severity, human=human, step=step,
                step_index=step_index, step_total=step_total, data=dict(data),
            )
            self._history.setdefault(run_id, deque(maxlen=self._capacity)).append(envelope)
            self._published.notify_all()
            return envelope

    def since(self, run_id: str, since_seq: int) -> tuple[list[EventEnvelope], int]:
        """``(events after since_seq, number dropped)``.

        A client that slept longer than the buffer is told how many events it will never see, rather
        than receiving a short replay that looks complete.
        """
        with self._lock:
            history = self._history.get(run_id)
            if not history:
                return [], 0
            oldest = history[0].seq
            dropped = max(0, oldest - (since_seq + 1))
            return [e for e in history if e.seq > since_seq], dropped

    def latest_seq(self, run_id: str) -> int:
        with self._lock:
            return self._next_seq.get(run_id, 1) - 1

    def wait_for(self, run_id: str, after_seq: int, *, timeout: float) -> list[EventEnvelope]:
        """Block until an event after ``after_seq`` exists, or the timeout expires.

        Returns an empty list on timeout, which is what lets a WebSocket loop send a keep-alive and
        then ask again: a socket that is merely quiet is indistinguishable from a dead one otherwise.
        """
        deadline = time.monotonic() + timeout
        with self._published:
            while True:
                pending = [
                    e for e in self._history.get(run_id, ()) if e.seq > after_seq
                ]
                if pending:
                    return pending
                remaining = deadline - time.monotonic()
                if remaining <= 0:
                    return []
                self._published.wait(remaining)

    def forget(self, run_id: str) -> None:
        """Drop a run's history. Sequence numbers are not reset: a re-used id would replay stale seqs."""
        with self._lock:
            self._history.pop(run_id, None)
